### PR TITLE
Cce analytic test linearized bondi sachs

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Add the analytic test for a linearized Bondi-Sachs solution.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/Systems/Cce/AnalyticSolutions/SphericalMetricData.cpp [hunk 1]
```diff
@@ -47,7 +47,7 @@ void SphericalMetricData::variables_impl(
       get<Tags::detail::InverseCartesianToSphericalJacobian>(
           intermediate_variables);
 
-  spherical_metric(make_not_null(&intermediate_spherical_metric), time);
+  spherical_metric(make_not_null(&intermediate_spherical_metric), l_max, time);
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/SphericalMetricData.cpp:49]: spelling error in commit message;  also you added a pup function which should be mentioned or a separate commit (and is it tested?)

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+ * transformations performed in the SpECTRE regularity-preserving CCE. Using the
+ * notation of \cite Barkett2019uae, we set:
+ *
+ * \f{align*}{
+ * B_2 &= B_3 = 0\\
+ * C_{2b} &= 3 C_{2a} / \nu^2\\
+ * C_{3b} &= -3 i C_{3a} / \nu^3
+ * \f}
+ *
+ * where \f$C_2a\f$ and \f$C_3a\f$ may be specified freely and are taken via
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp]: fix the superscripts

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+   * \f{align*}{
+   * \partial_r g_{a b} dx^a dx^b =& - (W + r \partial_r W
+   * - 2 r h_{A B} U^A U^B - r^2 (\partial_r h_{A B}) U^A U^B
+   * - 2 r^2 h_{A B} U^A \partial_r U^B) du^2 \\
+   * &- (4 r h_{A B} U^B + 2 r^2 ((\partial_r h_{A B}) U^B
+   * + h_{AB} \partial_r U^B) ) du dx^A
+   * + (2 r h_{A B} + r^2 \partial_r h_{A B}) dx^A dx^B,
+   * \f}
+   *
+   * where capital letters refer to angular coordinates and the angular tensors
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp]: do you mean indices with capital letters?

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+LinearizedBondiSachs::LinearizedBondiSachs(
+    const std::vector<std::complex<double>>& mode_constants,
+    const double extraction_radius, const double frequency) noexcept
+    : SphericalMetricData{extraction_radius}, frequency_{frequency} {
+  auto mode_constant_copy = mode_constants;
+  mode_constant_copy.resize(2, std::complex<double>(0.0, 0.0));
+  c_2a_ = gsl::at(mode_constant_copy, 0_st);
+  c_3a_ = gsl::at(mode_constant_copy, 1_st);
+  c_2b_ = 3.0 * c_2a_ / square(frequency_);
+  c_3b_ = -3.0 * std::complex<double>(0.0, 1.0) * c_3a_ / pow<3>(frequency_);
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: `std::complex<double>(0.0, -3.0)`

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+
+  const auto& collocation_metadata =
+      Spectral::Swsh::cached_collocation_metadata<
+          Spectral::Swsh::ComplexRepresentation::Interleaved>(l_max);
+  for (const auto& collocation_point : collocation_metadata) {
+    // assign to collocation points
+    const std::complex<double> z_22_factor =
+        (y_22.evaluate(collocation_point.theta, collocation_point.phi) +
+         y_2m2.evaluate(collocation_point.theta, collocation_point.phi)) /
+        sqrt(2.0);
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: why divide sqrt(2.0) here?  It makes it hard to compare with the documentation

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+    const size_t l_max, const double time) const noexcept {
+  const std::complex<double> j_2_coefficient =
+      0.5 * sqrt(6.0) *
+      (c_2a_ / extraction_radius_ - c_2b_ / (3.0 * pow<3>(extraction_radius_)));
+
+  const std::complex<double> j_3_coefficient =
+      2.0 * sqrt(15.0) *
+      (0.1 * c_3a_ / extraction_radius_ -
+       0.25 * c_3b_ / pow<4>(extraction_radius_) +
+       std::complex<double>(0.0, 1.0) * frequency_ *
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: use subtraction instead of adding a negative number...

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+
+void LinearizedBondiSachs::linearized_bondi_j(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, 2>*> bondi_j,
+    const size_t l_max, const double time) const noexcept {
+  const std::complex<double> j_2_coefficient =
+      0.5 * sqrt(6.0) *
+      (c_2a_ / extraction_radius_ - c_2b_ / (3.0 * pow<3>(extraction_radius_)));
+
+  const std::complex<double> j_3_coefficient =
+      2.0 * sqrt(15.0) *
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: multiply through by 2

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+}
+
+void LinearizedBondiSachs::linearized_bondi_u(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, 1>*> bondi_u,
+    const size_t l_max, const double time) const noexcept {
+  const std::complex<double> u_2_coefficient =
+      sqrt(6.0) * (0.5 * c_2a_ / square(extraction_radius_) +
+                   0.25 * c_2b_ / pow<4>(extraction_radius_) +
+                   std::complex<double>(1.0, 0.0) * frequency_ *
+                       (c_2b_ / (3.0 * extraction_radius_)));
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: missing cube of extraction_radius_?

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+                                   l_max, time);
+}
+
+void LinearizedBondiSachs::linearized_bondi_u(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, 1>*> bondi_u,
+    const size_t l_max, const double time) const noexcept {
+  const std::complex<double> u_2_coefficient =
+      sqrt(6.0) * (0.5 * c_2a_ / square(extraction_radius_) +
+                   0.25 * c_2b_ / pow<4>(extraction_radius_) +
+                   std::complex<double>(1.0, 0.0) * frequency_ *
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: this is an `std::complex<double>(0.0, 1.0)` in the documentation

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+      -square(frequency_) * c_2b_ / square(extraction_radius_) +
+      0.5 * c_2b_ / pow<4>(extraction_radius_) +
+      std::complex<double>(0.0, 1.0) * frequency_ *
+          (c_2b_ / pow<3>(extraction_radius_));
+
+  const std::complex<double> w_3_coefficient =
+      2.5 * frequency_ * c_3b_ / pow<4>(extraction_radius_) +
+      3.0 * c_3b_ / pow<5>(extraction_radius_) +
+      std::complex<double>(0.0, 1.0) * frequency_ *
+          (-2.0 * square(frequency_) * c_3b_ / square(extraction_radius_) -
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: this term is positive in the documentation

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+                                   dr_j_3_coefficient, l_max, time);
+}
+
+void LinearizedBondiSachs::linearized_dr_bondi_u(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, 1>*> dr_bondi_u,
+    const size_t l_max, const double time) const noexcept {
+  const std::complex<double> dr_u_2_coefficient =
+      sqrt(6.0) * (-c_2a_ / pow<3>(extraction_radius_) -
+                   c_2b_ / pow<5>(extraction_radius_) +
+                   std::complex<double>(1.0, 0.0) * frequency_ *
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: should be `i` not 1

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+                       (-c_2b_ / (3.0 * square(extraction_radius_))));
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: should be `pow<4>` not square

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 13]
```diff
@@ -0,0 +1,511 @@
...
+
+  const std::complex<double> dr_u_3_coefficient =
+      sqrt(12.0) *
+      (-c_3a_ / pow<3>(extraction_radius_) +
+       2.0 * square(frequency_) * c_3b_ / (pow<4>(extraction_radius_)) -
+       5.0 * c_3b_ / pow<6>(extraction_radius_) +
+       std::complex<double>(0.0, 1.0) * frequency_ *
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp:201]: you can put something besides 1.0 in the constructor of the a std::complex 😛

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+   *  \mathrm{Re}(\partial_r W_2(r) e^{i \nu u})
+   * + \frac{1}{\sqrt{2}} ({}_0 Y_{3\,3} - {}_0 Y_{3\, -3})
+   * \mathrm{Re}(\partial_r W_3(r) e^{i \nu u}),
+   * \f]
+   *
+   * where
+   *
+   * \f{align*}{
+   * \partial_r W_2(r) &= \frac{2 \nu^2 C_{2b}}{r^3}
+   * - \frac{3 i \nu C_{2 b}}{r^4} - \frac{2 C_2b}{r^5}, \\
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp]: fix subscript `C_2b`

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 15]
```diff
@@ -0,0 +1,511 @@
...
+      2.0 * square(frequency_) * c_2b_ / pow<3>(extraction_radius_) -
+      2.0 * c_2b_ / pow<5>(extraction_radius_) +
+      std::complex<double>(0.0, 1.0) * frequency_ *
+          (-3.0 * c_2b_ / pow<4>(extraction_radius_));
+
+  const std::complex<double> dr_w_3_coefficient =
+      -10.0 * frequency_ * c_3b_ / pow<5>(extraction_radius_) -
+      15.0 * c_3b_ / pow<6>(extraction_radius_) +
+      std::complex<double>(0.0, 1.0) * frequency_ *
+          (4.0 * square(frequency_) * c_3b_ / pow<3>(extraction_radius_) +
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: doc has -4.0

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 16]
```diff
@@ -0,0 +1,511 @@
...
+  const std::complex<double> dr_w_2_coefficient =
+      2.0 * square(frequency_) * c_2b_ / pow<3>(extraction_radius_) -
+      2.0 * c_2b_ / pow<5>(extraction_radius_) +
+      std::complex<double>(0.0, 1.0) * frequency_ *
+          (-3.0 * c_2b_ / pow<4>(extraction_radius_));
+
+  const std::complex<double> dr_w_3_coefficient =
+      -10.0 * frequency_ * c_3b_ / pow<5>(extraction_radius_) -
+      15.0 * c_3b_ / pow<6>(extraction_radius_) +
+      std::complex<double>(0.0, 1.0) * frequency_ *
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: factor out another factor of 4.0*frequency

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 17]
```diff
@@ -0,0 +1,511 @@
...
+          (4.0 * square(frequency_) * c_3b_ / pow<3>(extraction_radius_) +
+           12.0 * frequency_ * c_3b_ / pow<4>(extraction_radius_));
+
+  assign_components_from_l_factors(dr_bondi_w, dr_w_2_coefficient,
+                                   dr_w_3_coefficient, l_max, time);
+}
+
+void LinearizedBondiSachs::linearized_dt_bondi_j(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, 2>*> dt_bondi_j,
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: document that there is a factor of (we don't know) * `i` added in the `assign_dt_components_from_l_factors` call compared to the documentation...

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 18]
```diff
@@ -0,0 +1,511 @@
...
+                                      j_3_coefficient, l_max, time);
+}
+
+void LinearizedBondiSachs::linearized_dt_bondi_u(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, 1>*> dt_bondi_u,
+    const size_t l_max, const double time) const noexcept {
+  const std::complex<double> u_2_coefficient =
+      sqrt(3.0) * (0.5 * c_2a_ / square(extraction_radius_) +
+                   0.25 * c_2b_ / pow<4>(extraction_radius_) +
+                   std::complex<double>(1.0, 0.0) * frequency_ *
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: this should be `i`

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 19]
```diff
@@ -0,0 +1,511 @@
...
+                       (c_2b_ / (3.0 * extraction_radius_)));
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: cube of extraction radius

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 20]
```diff
@@ -0,0 +1,511 @@
...
+      -square(frequency_) * c_2b_ / square(extraction_radius_) +
+      0.5 * c_2b_ / pow<4>(extraction_radius_) +
+      std::complex<double>(0.0, 1.0) * frequency_ *
+          (c_2b_ / pow<3>(extraction_radius_));
+
+  const std::complex<double> w_3_coefficient =
+      2.5 * frequency_ * c_3b_ / pow<4>(extraction_radius_) +
+      3.0 * c_3b_ / pow<5>(extraction_radius_) +
+      std::complex<double>(0.0, 1.0) * frequency_ *
+          (-2.0 * square(frequency_) * c_3b_ / square(extraction_radius_) -
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: This is 2.0 in the documentation

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp [hunk 21]
```diff
@@ -0,0 +1,476 @@
...
+   *  \mathrm{Re}(W_2(r) e^{i \nu u})
+   * + \frac{1}{\sqrt{2}} i \nu ({}_1 Y_{3\,3} - {}_1 Y_{3\, -3})
+   * \mathrm{Re}(W_3(r) e^{i \nu u}),
+   * \f]
+   *
+   * where
+   *
+   * \f{align*}{
+   * W_2(r) &= - \frac{\nu^2 C_{2b}}{r^2} + \frac{i \nu C_{2 b}}{r^3}
+   * + \frac{C_2b}{2 r^4}, \\
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp]: fix subscript

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 22]
```diff
@@ -0,0 +1,511 @@
...
+  auto& bondi_u = get(get<Tags::BondiU>(temporary_variables));
+  auto& bondi_k = get(get<Tags::BondiK>(temporary_variables));
+  auto& bondi_w = get(get<Tags::BondiW>(temporary_variables));
+
+  linearized_bondi_j(make_not_null(&bondi_j), l_max, time);
+  linearized_bondi_u(make_not_null(&bondi_u), l_max, time);
+  linearized_bondi_w(make_not_null(&bondi_w), l_max, time);
+  bondi_k = sqrt(1.0 + bondi_j * conj(bondi_j));
+
+  get<0, 0>(*spherical_metric) =
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp:308]: we're too confused to compare with the documentation...

**Review by kidder — CHANGES_REQUESTED** [review]
reviewed with @nilsdeppe

File: src/Evolution/Systems/Cce/AnalyticSolutions/SphericalMetricData.cpp (hunk 1 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/SphericalMetricData.cpp:49]: The pup function for the `SphericalMetricData` class is tested indirectly by the test of the serialization of `LinearizedBondiSachs` (and will be also tested by other derived classes), but isn't tested directly because `SphericalMetricData` is abstract. Is that sufficient?

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp (hunk 17 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: Sorry, I'm a little confused - do you mean the `frequency_ * (std::complex<double>(0.0, 1.0) * cos(frequency_ * time) + sin(frequency_ * time))`? That's just equal to $i \nu e^{i \nu t}$ that I think made it into the documentation -- though it is in the wrong place, I'll move it inside the Re() operator. Was there another factor that's documented wrong?

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp (hunk 22 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp:308]: Oh, yeah, you're right -- I think the angular factors are probably too complicated to present as I have done. I'll just redo the documentation to explicitly list out the components.

File: src/Evolution/Systems/Cce/AnalyticSolutions/SphericalMetricData.cpp (hunk 1 above)
Comment (kidder) [reply src/Evolution/Systems/Cce/AnalyticSolutions/SphericalMetricData.cpp:49]: yes

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp (hunk 17 above)
Comment (kidder) [reply src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: there were factors of  `sqrt(2.0)` that we were confused about, that it looks like you've removed...

Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: oh, I see - yeah, there were some in the dt version as well that I didn't finish factoring out consistently. I added another fixup to make that all clear and correct. Thanks!

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp [hunk 23]
```diff
@@ -178,7 +178,7 @@ struct LinearizedBondiSachs : public SphericalMetricData {
    * \f{align*}{
    * W_2(r) &= - \frac{\nu^2 C_{2b}}{r^2} + \frac{i \nu C_{2 b}}{r^3}
    * + \frac{C_2b}{2 r^4}, \\
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp]: fix this subscript

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 24]
```diff
@@ -214,17 +214,19 @@ void LinearizedBondiSachs::linearized_dr_bondi_w(
...
-      15.0 * c_3b_ / pow<6>(extraction_radius_) +
-      std::complex<double>(0.0, 1.0) * frequency_ *
-          (4.0 * square(frequency_) * c_3b_ / pow<3>(extraction_radius_) +
-           12.0 * frequency_ * c_3b_ / pow<4>(extraction_radius_));
+      (-10.0 * frequency_ * c_3b_ / pow<5>(extraction_radius_) -
+       15.0 * c_3b_ / pow<6>(extraction_radius_) +
+       std::complex<double>(0.0, 4.0) * square(frequency_) * c_3b_ *
+           (frequency_ / pow<3>(extraction_radius_) +
+            3.0 / pow<4>(extraction_radius_))) /
+      sqrt(2.0);
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp:224]: we think dt_bondi_j is incorrect

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 25]
```diff
@@ -308,12 +310,10 @@ void LinearizedBondiSachs::spherical_metric(
...
-               square(bondi_u.data()) +
-           bondi_k.data() * square(extraction_radius_) * bondi_u.data() *
-               conj(bondi_u.data()) -
-           extraction_radius_ * bondi_w.data());
+      -real(1.0 + extraction_radius_ * bondi_w.data() -
+            square(extraction_radius_) *
+                (conj(bondi_j.data()) * square(bondi_u.data()) +
+                 bondi_k.data() * bondi_u.data() * conj(bondi_u.data())));
   get<0, 1>(*spherical_metric) = -1.0;
   get<0, 2>(*spherical_metric) =
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp:314]: please check (0,2) and (0,3) as they don't agree with the doc

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 26]
```diff
@@ -461,11 +461,11 @@ void LinearizedBondiSachs::dt_spherical_metric(
...
+      square(extraction_radius_) * (conj(dt_bondi_j.data()) * bondi_u.data() +
+                                    conj(bondi_j.data()) * dt_bondi_u.data());
   const ComplexDataVector dt_r_squared_k_ubar =
-      dt_bondi_k.data() * square(extraction_radius_) * conj(bondi_u.data()) +
-      bondi_k.data() * square(extraction_radius_) * conj(dt_bondi_u.data());
+      square(extraction_radius_) * (dt_bondi_k.data() * conj(bondi_u.data()) +
+                                    bondi_k.data() * conj(dt_bondi_u.data()));
   get<0, 2>(*dt_spherical_metric) =
       real(dt_r_squared_jbar_u + dt_r_squared_k_ubar);
   get<0, 3>(*dt_spherical_metric) =
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp:474]: check the minus sign with respect to the doc and factor out the minus sign of (2,3)

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 27]
```diff
@@ -385,13 +385,13 @@ void LinearizedBondiSachs::dr_spherical_metric(
...
+      square(extraction_radius_) * (conj(dr_bondi_j.data()) * bondi_u.data() +
+                                    conj(bondi_j.data()) * dr_bondi_u.data());
   const ComplexDataVector dr_r_squared_k_ubar =
-      dr_bondi_k.data() * square(extraction_radius_) * conj(bondi_u.data()) +
       2.0 * bondi_k.data() * extraction_radius_ * conj(bondi_u.data()) +
-      bondi_k.data() * square(extraction_radius_) * conj(dr_bondi_u.data());
+      square(extraction_radius_) * (dr_bondi_k.data() * conj(bondi_u.data()) +
+                                    bondi_k.data() * conj(dr_bondi_u.data()));
   get<0, 2>(*dr_spherical_metric) =
       real(dr_r_squared_jbar_u + dr_r_squared_k_ubar);
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp:396]: check the minus sign of (0,3) and factor out the minus sign of (2,3)

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp [hunk 28]
```diff
@@ -461,11 +461,11 @@ void LinearizedBondiSachs::dt_spherical_metric(
...
   const ComplexDataVector dt_r_squared_jbar_u =
-      conj(dt_bondi_j.data()) * square(extraction_radius_) * bondi_u.data() +
-      conj(bondi_j.data()) * square(extraction_radius_) * dt_bondi_u.data();
+      square(extraction_radius_) * (conj(dt_bondi_j.data()) * bondi_u.data() +
+                                    conj(bondi_j.data()) * dt_bondi_u.data());
   const ComplexDataVector dt_r_squared_k_ubar =
-      dt_bondi_k.data() * square(extraction_radius_) * conj(bondi_u.data()) +
-      bondi_k.data() * square(extraction_radius_) * conj(dt_bondi_u.data());
+      square(extraction_radius_) * (dt_bondi_k.data() * conj(bondi_u.data()) +
+                                    bondi_k.data() * conj(dt_bondi_u.data()));
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp]: check that the conjugation of things with respect to the doc

**Review by kidder — CHANGES_REQUESTED** [review]
please don't squash...  with @nilsdeppe

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp (hunk 24 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp:224]: Which part? If it's the sqrt prefactor, I've factored a 1/4 from the coefficients of the Cs giving sqrt(12/16)=sqrt(.75); however, I'm sure there's many other ways I could have messed up the math here.

Comment (kidder) [reply src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp:224]: ah, it is correct at the end, but not after the first of the two fix-ups, which is when we looked at it...

Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp:224]: ah, sorry -- I did probably leave it in an inconsistent state before the last sqrt(2) factoring commit.

**Comment by moxcodes** [comment]
posted a new set of fixups - this one I have `fixup: factoring and conjugation tweaks` and `fixup: dt to du mistake` (the latter because I forgot in the first attempt here that the derivatives with respect to r in the non-null frame need to be corrected from the null quantities).  I also added a fix for clang tidy items that seemed relevant from running locally, but it seems there are plenty more clang-tidy warnings thrown from the CI.

**Comment by nilsdeppe** [comment]
The remaining clang-tidy errors are fixed in #2239

File: src/NumericalAlgorithms/Spectral/SwshTransform.hpp [hunk 29]
```diff
@@ -536,8 +535,7 @@ struct InverseSwshTransform<tmpl::list<TransformTags...>, Representation> {
   // file and friend requirements
   template <ComplexRepresentation FriendRepresentation, int FriendSpin,
             typename... CollocationThenCoefficientTypes, size_t... Is>
-  // NOLINTNEXTLINE(readability-redundant-declaratioon)
-  friend void detail::inverse_swsh_transform_impl(
+  friend void detail::inverse_swsh_transform_impl(  // NOLINT
```
Comment (kidder) [inline src/NumericalAlgorithms/Spectral/SwshTransform.hpp]: is there now multiple warnings for `redundant-declaration`?  If so we can disable one is SetupClangTidy.cmake and keep the one we no-linted before

**Review by kidder — CHANGES_REQUESTED** [review]
Please squash everything

File: src/NumericalAlgorithms/Spectral/SwshTransform.hpp (hunk 29 above)
Comment (moxcodes) [reply src/NumericalAlgorithms/Spectral/SwshTransform.hpp]: Oh! it's good you pointed this out. I'd switched to the trailing `NOLINT` because the previous version stopped working, but now that you point it out I realize it's because the previous version had a typo, so I'll just revert the change apart from fixing the typo.

Comment (nilsdeppe) [reply src/NumericalAlgorithms/Spectral/SwshTransform.hpp]: If you look at line 41 of `SetupClangTidy.cmake` we ignore a bunch of warnings already because they're either duplicates, wrong, or have too many false positives. Could you add the duplicate warning to that list?

**Comment by moxcodes** [comment]
fixups squashed in.
Cheers!

File: cmake/SetupClangTidy.cmake [hunk 30]
```diff
@@ -32,13 +32,14 @@
...
+set(CLANG_TIDY_IGNORE_CHECKS "*,-cppcoreguidelines-no-malloc,-llvm-header-guard,-google-runtime-int,-readability-else-after-return,-misc-noexcept-move-constructor,-misc-unconventional-assign-operator,-cppcoreguidelines-c-copy-assignment-signature,-modernize-raw-string-literal,-hicpp-*,-android-*,-cert-err58-cpp,-google-default-arguments,-fuchsia-*,-performance-noexcept-move-constructor,-modernize-use-trailing-return-type,-cert-oop54-cpp,-misc-definitions-in-headers,-cppcoreguidelines-macro-usage,-cppcoreguidelines-avoid-magic-numbers,-modernize-use-nodiscard,-readability-magic-numbers,-bugprone-exception-escape,-cert-msc32-c,-misc-non-private-member-variables-in-classes,-cppcoreguidelines-avoid-c-arrays,-cppcoreguidelines-non-private-member-variables-in-classes,-cert-msc51-cpp,-bugprone-macro-parentheses,-readability-uppercase-literal-suffix,-readability-const-return-type,-readability-redundant-declaration")
```
Comment (nilsdeppe) [inline cmake/SetupClangTidy.cmake]: Now I'm even more confused. We've been ignorning the `readability-redundant-declaration` with friends because of the bug you noticed, but always using `NOLINTNEXTLINE`. However, the check is useful in general. In this PR it started showing up again and you had changed to a blanket `NOLINT`. The thing I'm not clear on is did the `NOLINTNEXTLINE` stop work, the name of the check change, or another check that does the same thing get added?

Comment (moxcodes) [reply cmake/SetupClangTidy.cmake]: Sorry, as I tried to clarify in my last comment, the `NOLINTNEXTLINE` previously (mysteriously) worked despite a spelling mistake in the error name that I only noticed after @kidder 's comment. I added the this to the cmake following your comment on the same thread, having misunderstood the request, so I'll just remove it now.

The change here will then just be correcting a spelling mistake in the `NOLINTNEXTLINE`

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

File: cmake/SetupClangTidy.cmake (hunk 30 above)
Comment (nilsdeppe) [reply cmake/SetupClangTidy.cmake]: Ah okay, sorry for the confusion!

**Review by nilsdeppe — APPROVED** [review]
